### PR TITLE
[6X] Fix query with DynamicBitmap scan crash when QueryFinishPending is true

### PR DESCRIPTION
**Pull request by yanwr1**
Previously, BitmapIndexScan create and free the bitmap memory. When BitmapHeapScan retrieves all tuples, it will free the bitmapstate's memory(tbmiterator, don't free bitmap), and the next relation's (of the same dynamic scan) BitmapHeapScan will first rescan the child node BitmapIndexScan which will free the bitmap memory.

But if QueryFinishPending is true when iteration over the bitmap, it will return null without free the bitmapstate, and next bitmapheap scan still need to do the rescan and free the bitmap memory. Finally when ExecEndBitmapHeapScan free bitmap iterator will access to a freed bitmap.

We don't have need to let BitmapIndexScan free bitmap memory, free bitmap iterator and bitmap memory together.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Review by chrishajas — APPROVED** [review]
Overall looks good with a couple minor comments.

This is a backport of https://github.com/greenplum-db/gpdb/pull/11408 right? Let's call out that we're backporting commit c270daec in the PR description and commit message, it makes it easier to look at the history and also review.

Also, it might be good to forward port the test case

File: src/backend/executor/nodeBitmapHeapscan.c [hunk 1]
```diff
@@ -51,6 +51,7 @@
 #include "utils/memutils.h"
```
Comment (dgkimura) [inline src/backend/executor/nodeBitmapHeapscan.c:51]: I noticed some diffs with the original patch.  Specifically, in `ExecEagerFreeBitmapHeapScan()` the order of close subplans and release bitmaps/buffers was swapped.  Any idea why was that required in 7X, but not in 6X?

On a related note - is the 7X GPDB comment [1] is out of date?

[1] https://github.com/greenplum-db/gpdb/blob/1aef04d/src/backend/executor/nodeBitmapHeapscan.c#L712-L722

Comment (yanwr1) [reply src/backend/executor/nodeBitmapHeapscan.c:51]: > I noticed some diffs with the original patch. Specifically, in `ExecEagerFreeBitmapHeapScan()` the order of close subplans and release bitmaps/buffers was swapped. Any idea why was that required in 7X, but not in 6X?
> in 6x, before this commit, bitmapIndexScan free the bitmap memory. when `ExecEagerFreeBitmapHeapScan()` free bitmap iterator still need to see bitmap's type, so have to release the iterators before closing down subplans. After the commit, i think the order have no impact.
> On a related note - is the 7X GPDB comment [1] is out of date?
> Yes, it's out of date. GP7 doesn't let bitmapIndexScan free bitmap memory now, close subplan won't release bitmap memory.

**Review by dgkimura — APPROVED** [review]
Looks good to me.

If you do forward port the test, it would be nice to fix up the outdated comment in `ExecEndBitmapHeapScan()` at the same time

File: src/backend/executor/nodeBitmapHeapscan.c [hunk 2]
```diff
@@ -668,6 +673,7 @@ BitmapAppendOnlyNext(BitmapHeapScanState *node)
 			return ExecClearTuple(slot);
 		}
 
+		node->tbm = tbm;
```
Comment (interma) [inline src/backend/executor/nodeBitmapHeapscan.c:676]: compared with the similar logic in `BitmapHeapNext()`:
https://github.com/greenplum-db/gpdb/blob/af903883053a14023064b4b4d28687c5f38ff86c/src/backend/executor/nodeBitmapHeapscan.c#L291-L300

Do we also need `node->tbmres = tbmres = NULL;` here?

Comment (interma) [reply src/backend/executor/nodeBitmapHeapscan.c:676]: And if `node->tbm` is always NULL here?
If so, let's add an `Assert()` here to ensure it.

Comment (yanwr1) [reply src/backend/executor/nodeBitmapHeapscan.c:676]: > Do we also need to set a proper value for `node->tbmres` here?

ExecInitBitmapHeapScanForPartition has set node->tbmres=NULL;

**Comment by yanwr1** [comment]
dev pipeline have been passed: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/dynamic

**Comment by interma** [comment]
> Looks like freeing both of them is a natural and simple behavior, so why did it only free one of them previously? Is there any hidden secret?

Have checked related code, looks it's ok. (no special purpose here)

**Review by interma — APPROVED** [review]
LGTM, thanks Wenru.

And there are some small differences between `BitmapAppendOnlyNext()` and `BitmapHeapNext()`, please help to check if has any risk here.

**Comment by kainwen** [comment]
@yanwr1 would you please help pre-port the test case here to main branhc?

**Comment by yanwr1** [comment]
> @yanwr1 would you please help pre-port the test case here to the main branch? Thanks.

ok, will do it later.
